### PR TITLE
apcu adapter

### PR DESCRIPTION
**Pull request by juliangut**
memcached adapter will have to wait till this gets merged

### CODE REVIEW

File: phpstan.neon.dist [hunk 1]
```diff
@@ -7,6 +7,8 @@ parameters:
     ignoreErrors:
         - '~Method [a-zA-Z0-9\\_]+::from[a-zA-Z0-9_]*\(\) has no return typehint specified.~'
         - '~Method [a-zA-Z0-9\\_]+Rate::[a-zA-Z0-9_]*\(\) has no return typehint specified.~'
+        - message: '~Strict comparison using === between int and false will always evaluate to false.~'
```
Comment (nikolaposa) [inline phpstan.neon.dist:10]: Where is this actually happening?

Comment (juliangut) [reply phpstan.neon.dist:10]: https://github.com/juliangut/rate-limit/blob/feature/apcu-adapter/src/ApcuRateLimiter.php#L84

**Comment by juliangut** [comment]
I think this PR is ready

**Comment by nikolaposa** [comment]
You should fix `composer.json` conflict.

I'll have more time to review PRs during the weekend.

**Comment by juliangut** [comment]
I didn't previously see the conflict. As mentioned above if this gets merged I can PR the memcached adapter and I'll be done with the original contribution

**Comment by Rainrider** [comment]
Please correct me if I'm wrong, but this basically doubles memory usage just to satisfy the constraints of the SilentRateLimiter interface, even if the user does not want to use it. Wouldn't it be better if `Status::$resetAt` has a possible null value when we can't get ttl values from the cache provider like it is the case with APCu?

**Comment by juliangut** [comment]
@Rainrider you are right, the amount of memory needed for this implementation, or any other in which remaining ttl is not returned by the provider, is more than doubled

There is another possible implementation in which you save both counter and timestamp in a single key: "counter::timestamp", but you loose atomic operations such as `apcu_inc` because you have to retrieve the value, explode, increment the counter and store it again, race conditions are more likely to occur

About not supporting remaining time to reset on an HTTP header I wouldn't consider it. If you have a peak to API rate limiting headers in the wild you'll see it is always there in a way or another (https://stackoverflow.com/questions/16022624/examples-of-http-api-rate-limiting-http-response-headers) there is even a IETF proposal for this headers (https://tools.ietf.org/id/draft-polli-ratelimit-headers-00.html), strange enough they are not standard

**Comment by nikolaposa** [comment]
I think @Rainrider brought up a good point. Why does default rate limiting implementation needs to be coupled with the logic around `$timeKey`? The very reason why interfaces are segregated is that the default implementation is intended for scenarios of having rate limit guard clauses in contexts that do not care about status information (i.e. remaining attempts), so it should have at less as possible communication with the back-end, APC in this case.

Try to focus on optimizing `limit()` and `limitSilently()` methods for their individual purposes, as if you are writing them in separate classes, without worrying about code reuse. It is completely ok that internally they have completely different flow of operation.

**Comment by nikolaposa** [comment]
Build is stuck on this one for 2 days... 🤔

**Comment by juliangut** [comment]
It's not stuck, if you take a look you'll see it passed, but github does not reflect it

File: src/ApcuRateLimiter.php [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+        $limitKey = $this->limitKey($identifier, $interval);
+        $timeKey = $this->timeKey($identifier, $interval);
+
+        $current = $this->getCurrent($limitKey);
+        if ($current <= $rate->getOperations()) {
+            $current = $this->updateCounterAndTime($limitKey, $timeKey, $interval);
+        }
+
+        return Status::from(
+            \sprintf('%s%s', $this->keyPrefix, $identifier),
```
Comment (nikolaposa) [inline src/ApcuRateLimiter.php]: `keyPrefix` is not needed here. `Status` is storage-agnostic, it provides rate limiting information about the original identifier (IP address, user ID, API key) regardless of whether information is stored in Redis, APC or some other backend. `keyPrefix` is an implementation detail of this particular backend.

File: src/ApcuRateLimiter.php [hunk 3]
```diff
@@ -0,0 +1,96 @@
...
+        }
+
+        $this->updateCounter($limitKey, $rate->getInterval());
+    }
+
+    public function limitSilently(string $identifier, Rate $rate): Status
+    {
+        $interval = $rate->getInterval();
+        $limitKey = $this->limitKey($identifier, $interval);
+        $timeKey = $this->timeKey($identifier, $interval);
```
Comment (nikolaposa) [inline src/ApcuRateLimiter.php:44]: Just got one (insufficiently tested 🙂 ) idea for building a key with embedded time information:
```
$key = sprintf(
    '%s:%s',
    $this->keyPrefix.$identifier,
    floor(time() / $rate->getInterval())
);
```

The idea with second part is the put identifiers into consistent time buckets. Therefore `Status->resetAt` boils down to: 
```
$resetAt = substr($key, strpos($key, ':') + 1) + $rate->getInterval();
```
right? 🤔

Comment (juliangut) [reply src/ApcuRateLimiter.php:44]: Although the idea of time buckets is good I'm afraid it's not feasible just like that, or I'm getting the idea wrong?

time() = 1590416425 (05/25/2020 @ 2:20pm (UTC))
for a 10 seconds interval, floor(1590416425 / 10) = 159041642
159041642 + 10 = 159041652 (01/15/1975 @ 6:14pm (UTC))

Comment (nikolaposa) [reply src/ApcuRateLimiter.php:44]: You're right, I was just thinking out loud. These buckets are useful for auto-expiry, calculating TTL is possible, but it's not easy/accurate.

**Pull request merged**